janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphTraversalUtil.java

```diff
@@ -54,6 +56,22 @@
     private static final List<Class<? extends TraversalParent>> MULTIQUERY_COMPATIBLE_STEPS =
             Arrays.asList(BranchStep.class, OptionalStep.class, RepeatStep.class, TraversalFilterStep.class);
 
+    public static StandardJanusGraph getJanusGraph(final Traversal.Admin<?, ?> traversal) {
```

Comment:
(nitpick) I also checked that `getGraph` returns a new `Optional` object rather than returning the same object. So, maybe it's better to reduce the invocation of `getGraph` by storing the `Optional` object into a local variable. 
I made a quick check that your bytecode of this method uses 84 instructions. I tried to rewrite the method with `do while` and the bytecode changed to 76 instructions. That said, I didn't check the method correctness and didn't analyze the bytecode, so it doesn't mean anything. I don't say that you need to rewrite this method. It's up to you. I would probably store Optional graph somewhere like `Optional<Graph> optionalGraph = traversal.getGraph();` but that's not necessary.
The re-written method (which I didn't check): 
```
public static StandardJanusGraph getJanusGraph(Traversal.Admin<?, ?> traversal) {

    do {

        Optional<Graph> optionalGraph = traversal.getGraph();
        if (!optionalGraph.isPresent()) {
            return null;
        }

        Graph graph = optionalGraph.get();

        if(graph instanceof StandardJanusGraph){
            return (StandardJanusGraph) graph;
        }

        if(graph instanceof StandardJanusGraphTx){
            return ((StandardJanusGraphTx) graph).getGraph();
        }

        if (traversal.getParent() instanceof EmptyStep) {
            return null;
        }

        traversal = traversal.getParent().asStep().getTraversal();

    } while (true);
}
```

It's bytecode (which I didn't check):
```
public static getJanusGraph(Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin;)Lorg/janusgraph/graphdb/database/StandardJanusGraph;
   L0
    LINENUMBER 63 L0
   FRAME SAME
    ALOAD 0
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getGraph ()Ljava/util/Optional; (itf)
    ASTORE 1
   L1
    LINENUMBER 64 L1
    ALOAD 1
    INVOKEVIRTUAL java/util/Optional.isPresent ()Z
    IFNE L2
   L3
    LINENUMBER 65 L3
    ACONST_NULL
    ARETURN
   L2
    LINENUMBER 68 L2
   FRAME APPEND [java/util/Optional]
    ALOAD 1
    INVOKEVIRTUAL java/util/Optional.get ()Ljava/lang/Object;
    CHECKCAST org/apache/tinkerpop/gremlin/structure/Graph
    ASTORE 2
   L4
    LINENUMBER 70 L4
    ALOAD 2
    INSTANCEOF org/janusgraph/graphdb/database/StandardJanusGraph
    IFEQ L5
   L6
    LINENUMBER 71 L6
    ALOAD 2
    CHECKCAST org/janusgraph/graphdb/database/StandardJanusGraph
    ARETURN
   L5
    LINENUMBER 74 L5
   FRAME APPEND [org/apache/tinkerpop/gremlin/structure/Graph]
    ALOAD 2
    INSTANCEOF org/janusgraph/graphdb/transaction/StandardJanusGraphTx
    IFEQ L7
   L8
    LINENUMBER 75 L8
    ALOAD 2
    CHECKCAST org/janusgraph/graphdb/transaction/StandardJanusGraphTx
    INVOKEVIRTUAL org/janusgraph/graphdb/transaction/StandardJanusGraphTx.getGraph ()Lorg/janusgraph/graphdb/database/StandardJanusGraph;
    ARETURN
   L7
    LINENUMBER 78 L7
   FRAME SAME
    ALOAD 0
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getParent ()Lorg/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent; (itf)
    INSTANCEOF org/apache/tinkerpop/gremlin/process/traversal/step/util/EmptyStep
    IFEQ L9
   L10
    LINENUMBER 79 L10
    ACONST_NULL
    ARETURN
   L9
    LINENUMBER 82 L9
   FRAME SAME
    ALOAD 0
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getParent ()Lorg/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent; (itf)
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent.asStep ()Lorg/apache/tinkerpop/gremlin/process/traversal/Step; (itf)
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Step.getTraversal ()Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin; (itf)
    ASTORE 0
   L11
    LINENUMBER 84 L11
    GOTO L0
   L12
    LOCALVARIABLE optionalGraph Ljava/util/Optional; L1 L11 1
    // signature Ljava/util/Optional<Lorg/apache/tinkerpop/gremlin/structure/Graph;>;
    // declaration: optionalGraph extends java.util.Optional<org.apache.tinkerpop.gremlin.structure.Graph>
    LOCALVARIABLE graph Lorg/apache/tinkerpop/gremlin/structure/Graph; L4 L11 2
    LOCALVARIABLE traversal Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin; L0 L12 0
    // signature Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin<**>;
    // declaration: traversal extends org.apache.tinkerpop.gremlin.process.traversal.Traversal$Admin<?, ?>
    MAXSTACK = 1
    MAXLOCALS = 3
```

Your current method's bytecode (which again I didn't check):
```
public static getJanusGraph(Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin;)Lorg/janusgraph/graphdb/database/StandardJanusGraph;
   L0
    LINENUMBER 60 L0
    ALOAD 0
    ASTORE 1
   L1
    LINENUMBER 61 L1
    ALOAD 0
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getGraph ()Ljava/util/Optional; (itf)
    INVOKEVIRTUAL java/util/Optional.get ()Ljava/lang/Object;
    CHECKCAST org/apache/tinkerpop/gremlin/structure/Graph
    ASTORE 2
   L2
    LINENUMBER 62 L2
   FRAME APPEND [org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin org/apache/tinkerpop/gremlin/structure/Graph]
    ALOAD 2
    INSTANCEOF org/janusgraph/graphdb/database/StandardJanusGraph
    IFNE L3
    ALOAD 2
    INSTANCEOF org/janusgraph/graphdb/transaction/StandardJanusGraphTx
    IFNE L3
   L4
    LINENUMBER 63 L4
    ALOAD 1
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getParent ()Lorg/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent; (itf)
    INSTANCEOF org/apache/tinkerpop/gremlin/process/traversal/step/util/EmptyStep
    IFEQ L5
   L6
    LINENUMBER 64 L6
    ACONST_NULL
    ARETURN
   L5
    LINENUMBER 66 L5
   FRAME SAME
    ALOAD 1
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getParent ()Lorg/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent; (itf)
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/step/TraversalParent.asStep ()Lorg/apache/tinkerpop/gremlin/process/traversal/Step; (itf)
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Step.getTraversal ()Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin; (itf)
    ASTORE 1
   L7
    LINENUMBER 67 L7
    ALOAD 1
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getGraph ()Ljava/util/Optional; (itf)
    INVOKEVIRTUAL java/util/Optional.isPresent ()Z
    IFNE L8
   L9
    LINENUMBER 68 L9
    ACONST_NULL
    ARETURN
   L8
    LINENUMBER 70 L8
   FRAME SAME
    ALOAD 1
    INVOKEINTERFACE org/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin.getGraph ()Ljava/util/Optional; (itf)
    INVOKEVIRTUAL java/util/Optional.get ()Ljava/lang/Object;
    CHECKCAST org/apache/tinkerpop/gremlin/structure/Graph
    ASTORE 2
    GOTO L2
   L3
    LINENUMBER 72 L3
   FRAME SAME
    ALOAD 2
    INSTANCEOF org/janusgraph/graphdb/transaction/StandardJanusGraphTx
    IFEQ L10
    ALOAD 2
    CHECKCAST org/janusgraph/graphdb/transaction/StandardJanusGraphTx
    INVOKEVIRTUAL org/janusgraph/graphdb/transaction/StandardJanusGraphTx.getGraph ()Lorg/janusgraph/graphdb/database/StandardJanusGraph;
    GOTO L11
   L10
   FRAME SAME
    ALOAD 2
    CHECKCAST org/janusgraph/graphdb/database/StandardJanusGraph
   L11
   FRAME SAME1 org/janusgraph/graphdb/database/StandardJanusGraph
    ARETURN
   L12
    LOCALVARIABLE traversal Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin; L0 L12 0
    // signature Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin<**>;
    // declaration: traversal extends org.apache.tinkerpop.gremlin.process.traversal.Traversal$Admin<?, ?>
    LOCALVARIABLE t Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin; L1 L12 1
    // signature Lorg/apache/tinkerpop/gremlin/process/traversal/Traversal$Admin<**>;
    // declaration: t extends org.apache.tinkerpop.gremlin.process.traversal.Traversal$Admin<?, ?>
    LOCALVARIABLE graph Lorg/apache/tinkerpop/gremlin/structure/Graph; L2 L12 2
    MAXSTACK = 1
    MAXLOCALS = 3
```